app/Http/Controllers/DeviceController.php

```diff
@@ -150,6 +150,19 @@ private function buildDeviceGraphArrays($device)
         return $graphs;
     }
 
+    private function buildDeviceLinkTargetValue(Device $device): string
```

Comment:
Hi
This private function is repeated twice. Probably better to write it once and use it twice instead.

Reply:
Yes, I wasn’t exactly happy with that, too, but that method is used across two different classes, both having only Controller as parent class. I don’t think that this class is a fitting place for that rather specific method, isn’t it? Do you have some suggestion where to place it? Sth. like a helper module/class/file?

Reply:
There is indeed such helpers : You can find them in `/opt/librenms/LibreNMS/Util`. And one particular class would be the perfect candidate : 
`/opt/librenms/LibreNMS/Util/Url.php`.

The function itself could then be : 
`public static function directDeviceUrl($device): string

And this function could be called from your code to build the HTTPs, telnet etc etc.  The class is already loaded in the 2 files (`use LibreNMS\Util\Url;`) so you can use any function of this class directly with something like :
`$telnet = 'telnet://' . Url::directDeviceUrl($device);`

Reply:
Thank you very much for your advice; apparently I was not able to detect that "Url" class import. Now, after being pointed out by you, it seems pretty obvious …
Nevertheless: I added a "Part" to the method name, because it is does not include the protocol, so Url alone might be misleading.